Average effectivePressure over hydrology subcycle

alexolinhager opened the pull request:
This PR defines a new variable, effectivePressureSGH, that is identical to the previous effectivePressure variable. effectivePressureSGH is calculated every hydrology timestep and is only used by the hydrology model. effectivePressure is now redefined as the average effectivePressureSGH throughout a dynamics coupling interval, and is only used by the dynamics model when running in a coupled configuration. The purpose of this PR is to increase stability when coupling the hydrology and dynamics models by making the dynamics model respond to the effectivePressure field throughout the hydrology subcycle, instead of an arbitrary sampling at its last timestep. This reduces the likelihood of pressure waves building resonance between the two model components.

matthewhoffman commented on components/mpas-albany-landice/src/Registry_subglacial_hydro.xml:
```diff
@@ -178,8 +178,10 @@
+                <var name="effectivePressureSGH" type="real" dimensions="nCells Time" units="Pa"
+                        description="effective ice pressure in subglacial hydrology system" />
+                <var name="effectivePressure" type="real" dimensions"nCells Time" units="Pa" packages="higherOrderVelocity"
+                        description="effectivePressureSGH averaged over timestepping interval. This is the version of effective pressure that is passed to the coupled dynamics model."/>
```
```suggestion
                        description="Effective pressure used in basal friction calculation.  If subglacial hydrology model is active, this will be effectivePressureSGH averaged over the subglacial hydrology model timestepping subcycles. 
 If subglacial hydrology model in inactive, this will come from a file or a parameterization."/>
```

matthewhoffman commented on components/mpas-albany-landice/src/Registry_subglacial_hydro.xml:
```diff
@@ -178,8 +178,10 @@
-                     description="effective ice pressure in subglacial hydrology system" />
+                <var name="effectivePressureSGH" type="real" dimensions="nCells Time" units="Pa"
+                        description="effective ice pressure in subglacial hydrology system" />
+                <var name="effectivePressure" type="real" dimensions"nCells Time" units="Pa" packages="higherOrderVelocity"
```
Won't compile with this missing =
```suggestion
                <var name="effectivePressure" type="real" dimensions="nCells Time" units="Pa" packages="higherOrderVelocity"
```

matthewhoffman commented on components/mpas-albany-landice/src/mode_forward/mpas_li_subglacial_hydro.F:
```diff
@@ -2229,7 +2256,7 @@ subroutine ocean_connection_N(domain)
       type (mpas_pool_type), pointer :: geometryPool
       real (kind=RKIND), dimension(:), pointer :: bedTopography
-      real (kind=RKIND), dimension(:), pointer :: effectivePressure
+      real (kind=RKIND), dimension(:), pointer :: effectivePressureSGH
```
For the entire `ocean_connection_N` subroutine, we want to keep the variable as `effectivePressure`

matthewhoffman replied:
@alexolinhager , I'm still seeing `effectivePressureSGH` being used in the `ocean_connection_N` subroutine.

alexolinhager replied:
Sorry about that, I had previously removed this but then reverted to do some debugging. I forgot to undo the revert when I realized this wasn't causing the issue. I just made this change, pushed the commit, and retesting the branch. Everything passes the full integration test

matthewhoffman reviewed (changes requested):
@alexolinhager , this looks great and is probably long overdue.  I have requested a few changes that will be quick to take care of.

alexolinhager commented:
@matthewhoffman, I applied your suggestions and also added a bug fix (commit [5a1aea6](https://github.com/MALI-Dev/E3SM/pull/122/commits/5a1aea6b6f9661a30ef5a80bf3cf9783e9ae015e) that I had implemented earlier but had forgot to push before you reviewed. The previous version was missing a `deltatSGH` that needs to be multiplied with `effectivePressure` in the weighted average calculation. Without it, velocities in the thwaites coupled runs are a few orders of magnitude too high, but with this term, everything behaves as expected. The example I had been working from (https://github.com/E3SM-Project/E3SM/blob/f50d2639c68c1607a521407b420af040080c1380/components/mpas-ocean/src/shared/mpas_ocn_time_average_coupled.F#L332-L333) has an implicit `*1` after `gradSSHZonal(iCell)` that gets dropped.

Unfortunately, the updated code now fails the baseline comparison for the Greenland fo decomposition/restart and Greenland sia decomposition/restart tests, which is weird because these don't have active hydrology. My original thought was the issue had to do with [6b20a2b](https://github.com/MALI-Dev/E3SM/pull/122/commits/6b20a2bd41063c7cb8bbef03572271d2977ab362), but I reverted that commit and the tests were the same. Is there anything in the updated code that you think could be causing those tests to fail?

alexolinhager commented:
![Screenshot 2024-08-29 at 1 15 05 PM](https://github.com/user-attachments/assets/a8480ef9-5229-4e8c-be87-ddfac2b83e29)

matthewhoffman commented on components/mpas-albany-landice/src/mode_forward/mpas_li_subglacial_hydro.F:
```diff
@@ -700,16 +702,24 @@ subroutine li_SGH_solve(domain, err)
+         if (deltatSGHaccumulated == 0.0_RKIND) then
+            effectivePressure = effectivePressureSGH
+         else
+            effectivePressure = (effectivePressure * deltatSGHaccumulated + effectivePressureSGH * deltatSGH) / (deltatSGHaccumulated + deltatSGH)
```
Good catch on correcting this - I didn't notice the missing factor.  However, no need for a special `(deltatSGHaccumulated == 0.0_RKIND)` case - line 714 will reduce to that already

matthewhoffman commented on components/mpas-albany-landice/src/mode_forward/mpas_li_subglacial_hydro.F:
```diff
@@ -700,16 +702,24 @@ subroutine li_SGH_solve(domain, err)
+            effectivePressure = (effectivePressure * deltatSGHaccumulated + effectivePressureSGH * deltatSGH) / (deltatSGHaccumulated + deltatSGH)
+         endif
+
+         dt = deltatSGH
```
I don't understand why a new `dt` variable was introduced - I don't think it's needed.

alexolinhager replied:
I had originally put this in because I wasn't sure how to use `deltatSGH` outside of the block, if I need `block % structs` to access the `hydroPool`

alexolinhager replied:
Could `deltatSGHaccumulated = deltatSGHaccumulated + deltatSGH` be placed inside the block? I wanted to make sure this counter wouldn't get activated by moving through blocks (if more than 1 exists)

alexolinhager replied:
Nevermind, I guess that's what we originally did, so I'm just going to stick to that original implementation

matthewhoffman replied:
I think the best solution is a combination of the two versions.  You are right that it would be better to accumulate outside the block, even though we don't support multiple blocks anymore.  You can still access `deltatSGH` outside the block - it will just retain the value from the final block, which would be ok in this case because all blocks should have the same value (if one hypothetically tried to run with multiple blocks).

alexolinhager commented:
@matthewhoffman, I cleaned up the commit history and everything passes the integration test now. I am happy with this getting merged if you are

<img width="1177" alt="Screenshot 2024-08-29 at 5 03 35 PM" src="https://github.com/user-attachments/assets/4f59ae16-a7e8-4dbe-8bca-a8bb4d3f4020">

matthewhoffman reviewed (changes requested):
@alexolinhager , thanks for revising this PR.  Some changes are still needed in the ocean_connection_N subroutine.
